crates/bevy_picking_selection/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+//!
+//! This adds the [`PointerDeselect`] and [`PointerSelect`] [`PointerEvent`]s, including support for
+//! bubbling these events.
+
+#![allow(clippy::type_complexity)]
+#![allow(clippy::too_many_arguments)]
+#![deny(missing_docs)]
+
+use bevy::prelude::*;
+use bevy_picking_core::{output::PointerEvent, pointer::PointerId, PickStage};
+
+/// Adds multiselect picking support to your app.
+pub struct SelectionPlugin;
+impl Plugin for SelectionPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_event::<PointerSelect>()
+            .add_event::<PointerDeselect>()
+            .add_system_set_to_stage(
+                CoreStage::First,
+                SystemSet::new()
+                    .after(PickStage::Focus)
+                    .before(PickStage::EventListeners)
+                    .with_system(multiselect_events)
+                    .with_system(send_selection_events.after(multiselect_events))
+                    .with_system(update_state_from_events.after(send_selection_events)),
+            )
+            .add_system_set_to_stage(
+                CoreStage::First,
+                SystemSet::new()
+                    .after(PickStage::Focus)
+                    .label(PickStage::EventListeners)
+                    .with_system(bevy_picking_core::output::event_bubbling::<Select>)
+                    .with_system(bevy_picking_core::output::event_bubbling::<Deselect>),
+            );
+    }
+}
+
+/// Input state that defines whether or not the multiselect button is active. This is often the
+/// `Ctrl` or `Shift` keys.
+#[derive(Debug, Default, Clone, Component, PartialEq)]
+pub struct PointerMultiselect {
+    /// `true` if the multiselect button(s) is active.
+    pub is_pressed: bool,
+}
+
+/// Tracks the current selection state of the entity.
+#[derive(Component, Debug, Default, Clone)]
+pub struct PickSelection {
+    /// `true` if this entity is selected.
+    pub is_selected: bool,
+}
+
+/// Fires when an entity has been selected
+pub type PointerSelect = PointerEvent<Select>;
+/// The inner [`PointerEvent`] type for [`PointerSelect`].
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Reflect)]
+pub struct Select;
+
+/// Fires when an entity has been deselected
+pub type PointerDeselect = PointerEvent<Deselect>;
+/// The inner [`PointerEvent`] type for [`PointerDeselect`].
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Reflect)]
+pub struct Deselect;
+
+/// Marker struct used to mark pickable entities for which you don't want to trigger a deselection
+/// event when picked. This is useful for gizmos or other pickable UI entities.
+#[derive(Component, Debug, Copy, Clone)]
+pub struct NoDeselect;
+
+/// Unsurprising default multiselect inputs: both  control and shift keys.
+pub fn multiselect_events(
+    keyboard: Res<Input<KeyCode>>,
+    mut pointer_query: Query<&mut PointerMultiselect>,
+) {
+    let is_multiselect_pressed = keyboard.any_pressed([
```

Comment:
How would a user provide custom multiselect_pressed KeyCodes?

Reply:
The same way it's provided in this system, they would just write their own version. How do you suggest this is communicated?

Reply:
Would the user have to write they own Selection plugin then? Would it make sense to have a way to specify the multislect keys when adding the plugin? Sorry if this doesn’t make sense, it’s really hot where I am so my brain is functioning poorly 😅

Reply:
They can just reuse the selection plugin, but replace this system with their own. :)

You raise a good point though, it would be a good idea to extract this out as its own function so it's even easier to disable.

Reply:
Good point.

The ability to disable is good, though I’m thinking that making multiselect configurable as well as disable-able would be ideal. I’m not sure how this would be accomplished, perhaps @alice-i-cecile has some insight as this problem reminds my of her Leafwing Input Manager’s ‘action binding’ solution.

Reply:
All inputs (even pointer clicks and movement) are exposed in this refactor, and the default bindings are easy to omit. This was done intentionally to support rebinding. This might make for a good example.

Reply:
Yeah; I'd be tempted to do a LWIM + bevy_mod_picking rebinding example. dev-dependencies are great for that.

Reply:
> All inputs (even pointer clicks and movement) are exposed in this refactor, and the default bindings are easy to omit. This was done intentionally to support rebinding. This might make for a good example.

Awesome. Thanks for the clarification.

Reply:
Could keybinds be supplied in a plugin configuration resource somewhere? That you expose what _actions_ can be taken and supply a default keybind that the user can override if they want to bind a different key to an action?

Reply:
> Could keybinds be supplied in a plugin configuration resource somewhere? 
I'm hesitant to do this because I don't want to make any assumptions about how inputs work. E.g. I don't want to constrain an input to only work with the keyboard or mouse. That's why all input events are just bevy events or resources, to allow users to update these inputs with whatever they want (gamepads, joysticks, the current temperature in Belize, a web request, etc. :) )